File: frontend/src/utils/trickUtils.ts
/**
 * trickUtils.ts — Trick resolution and move validation.
 * 
 * Contains trick winner determination, move validity checks, and
 * invalid move reason explanations (for Qayd disputes).
 */
import { CardModel, Rank, Suit, TableCardMetadata } from "../types";
import { getCardStrength } from "./scoringUtils";

export const getTrickWinner = (
    tableCards: { card: CardModel, playedBy: string, metadata?: TableCardMetadata }[],
    mode: 'SUN' | 'HOKUM',
    trumpSuit: Suit | null
): number => {
    if (tableCards.length === 0) return -1;

    const leadSuit = tableCards[0].card.suit;
    let highestStrength = -1;
    let winnerIndex = 0;

    tableCards.forEach((play, index) => {
        const card = play.card;
        let strength = 0;

        if (card.suit === leadSuit || (mode === 'HOKUM' && card.suit === trumpSuit)) {
            strength = getCardStrength(card, mode, trumpSuit);
        } else {
            strength = -1;
        }

        if (strength > highestStrength) {
            highestStrength = strength;
            winnerIndex = index;
        }
    });

    return winnerIndex;
};

export const isValidMove = (
    card: CardModel,
    hand: CardModel[],
    tableCards: { card: CardModel, metadata?: TableCardMetadata }[],
    mode: 'SUN' | 'HOKUM',
    trumpSuit: Suit | null,
    isLocked: boolean = false,
    strictMode: boolean = true
): boolean => {
    if (!strictMode) return true;

    // Lead player
    if (tableCards.length === 0) {
        if (isLocked && mode === 'HOKUM' && trumpSuit && card.suit === trumpSuit) {
            const hasNonTrump = hand.some(c => c.suit !== trumpSuit);
            if (hasNonTrump) return false;
        }
        return true;
    }

    const leadSuit = tableCards[0].card.suit;
    const hasLeadSuit = hand.some(c => c.suit === leadSuit);

    // Rule 1: Must follow suit
    if (hasLeadSuit) {
        if (card.suit !== leadSuit) return false;
        return true;
    }

    // Rule 2: Must cut with trump in Hokum if possible
    if (mode === 'HOKUM' && trumpSuit) {
        const hasTrump = hand.some(c => c.suit === trumpSuit);
        if (hasTrump && card.suit !== trumpSuit) return false;
    }

    return true;
};

/**
 * Explain WHY a move is invalid (for Disputes/Qayd).
 */
export const getInvalidMoveReason = (
    card: CardModel,
    hand: CardModel[],
    tableCards: { card: CardModel, metadata?: TableCardMetadata }[],
    mode: 'SUN' | 'HOKUM',
    trumpSuit: Suit | null,
    isLocked: boolean = false
): string | null => {
    if (tableCards.length === 0) {
        if (isLocked && mode === 'HOKUM' && trumpSuit && card.suit === trumpSuit) {
            const hasNonTrump = hand.some(c => c.suit !== trumpSuit);
            if (hasNonTrump) return "You cannot help (lead trump) when the game is Locked (Doubled)!";
        }
        return null;
    }

    const leadSuit = tableCards[0].card.suit;
    const hasLeadSuit = hand.some(c => c.suit === leadSuit);

    if (hasLeadSuit) {
        if (card.suit !== leadSuit) return `You have ${leadSuit} in your hand! You must follow suit (Renounce).`;
        return null;
    }

    if (mode === 'HOKUM' && trumpSuit) {
        const hasTrump = hand.some(c => c.suit === trumpSuit);
        if (hasTrump && card.suit !== trumpSuit) return `You have Trump (${trumpSuit})! You must Cut the trick since you cannot follow suit.`;
    }

    return null;
};

File: mobile/lib/utils/trick_utils.dart
/// trickUtils.dart — Trick resolution and move validation.
///
/// Port of frontend/src/utils/trickUtils.ts
import '../models/card_model.dart';
import '../models/enums.dart';
import 'scoring_utils.dart';

/// Determine the winner of a trick.
/// Returns the index (0-3) of the winning card in [tableCards].
int getTrickWinner(
  List<TableCard> tableCards,
  GameMode mode,
  Suit? trumpSuit,
) {
  if (tableCards.isEmpty) return -1;

  final leadSuit = tableCards[0].card.suit;
  int highestStrength = -1;
  int winnerIndex = 0;

  for (int i = 0; i < tableCards.length; i++) {
    final card = tableCards[i].card;
    int strength;

    if (card.suit == leadSuit ||
        (mode == GameMode.hokum && card.suit == trumpSuit)) {
      strength = getCardStrength(card, mode, trumpSuit);
    } else {
      strength = -1;
    }

    if (strength > highestStrength) {
      highestStrength = strength;
      winnerIndex = i;
    }
  }

  return winnerIndex;
}

/// Check if a card is a valid move.
bool isValidMove({
  required CardModel card,
  required List<CardModel> hand,
  required List<TableCard> tableCards,
  required GameMode mode,
  Suit? trumpSuit,
  bool isLocked = false,
  bool strictMode = true,
}) {
  // Null safety
  if (hand.isEmpty) return false;
  if (!strictMode) return true;

  // Lead player
  if (tableCards.isEmpty) {
    if (isLocked && mode == GameMode.hokum && trumpSuit != null && card.suit == trumpSuit) {
      final hasNonTrump = hand.any((c) => c.suit != trumpSuit);
      if (hasNonTrump) return false;
    }
    return true;
  }

  final leadSuit = tableCards[0].card.suit;
  final hasLeadSuit = hand.any((c) => c.suit == leadSuit);

  // Rule 1: Must follow suit
  if (hasLeadSuit) {
    return card.suit == leadSuit;
  }

  // Rule 2: Must cut with trump in Hokum if possible
  if (mode == GameMode.hokum && trumpSuit != null) {
    final hasTrump = hand.any((c) => c.suit == trumpSuit);
    if (hasTrump && card.suit != trumpSuit) return false;
  }

  return true;
}

/// Explain why a move is invalid (for Qayd disputes).
String? getInvalidMoveReason({
  required CardModel card,
  required List<CardModel> hand,
  required List<TableCard> tableCards,
  required GameMode mode,
  Suit? trumpSuit,
  bool isLocked = false,
}) {
  if (tableCards.isEmpty) {
    if (isLocked && mode == GameMode.hokum && trumpSuit != null && card.suit == trumpSuit) {
      final hasNonTrump = hand.any((c) => c.suit != trumpSuit);
      if (hasNonTrump) {
        return 'You cannot help (lead trump) when the game is Locked (Doubled)!';
      }
    }
    return null;
  }

  final leadSuit = tableCards[0].card.suit;
  final hasLeadSuit = hand.any((c) => c.suit == leadSuit);

  if (hasLeadSuit) {
    if (card.suit != leadSuit) {
      return 'You have ${leadSuit.symbol} in your hand! You must follow suit (Renounce).';
    }
    return null;
  }

  if (mode == GameMode.hokum && trumpSuit != null) {
    final hasTrump = hand.any((c) => c.suit == trumpSuit);
    if (hasTrump && card.suit != trumpSuit) {
      return 'You have Trump (${trumpSuit.symbol})! You must Cut the trick since you cannot follow suit.';
    }
  }

  return null;
}
